File: point/point.cpp
#include <cmath>
#include <iostream>
#include <memory>
#include <algorithm>

#include "point.hpp"
#include "utils.hpp"

Point::Point(double x, double y)
{
    this->_x = x;
    this->_y = y;
}

Point::~Point() = default;

double Point::x() const
{
    return this->_x;
}

double Point::y() const
{
    return this->_y;
}

double Point::distance(const Point& p, const Point& q)
{
    double dx = p.x() - q.x();
    double dy = p.y() - q.y();

    return sqrt(dx * dx + dy * dy);
}

File: go_point/point.go
package point

import "math"

type Point struct {
	x float64
	y float64
}

func NewPoint(pt *Point, x float64, y float64) *Point {
	if pt == nil {
		pt = &Point{}
	}

	pt.x = x
	pt.y = y

	return pt
}

func (pt Point) X() float64 {
	return pt.x
}

func (pt Point) Y() float64 {
	return pt.y
}

func DeletePoint(pt *Point) {}

func Distance(p *Point, q *Point) float64 {
	dx := p.X() - q.X()
	dy := p.Y() - q.Y()

	return math.Sqrt(dx*dx + dy*dy)
}
